### ISSUE TITLE
fix microversion header

### ISSUE THREAD

**Issue by jrperritt** [gophercloud/gophercloud#53]
Currently all `ServiceClient` methods erroneously add the Nova microversion header: https://github.com/gophercloud/gophercloud/blob/master/service_client.go#L52

Currently, I don't think there's a good way to know what type of `ServiceClient` the user has. One option would be to add a `Type` field to `ServiceClient` and switch on it in the (not yet existing) `ServiceClient` `Request` method.

**Comment by pospispa** [comment]
Created PR https://github.com/gophercloud/gophercloud/pull/341 to fix this issue.

**Comment by jtopjian** [comment]
Some non-coded thoughts on implementing this:

This might be relevant:

http://docs-draft.openstack.org/15/462815/4/check/gate-api-wg-docs-ubuntu-xenial/b8479be//doc/build/html/guidelines/consuming-catalog.html

As might this:

https://github.com/openstack/api-wg/blob/master/guidelines/microversion_specification.rst

> Currently, I don't think there's a good way to know what type of ServiceClient the user has. One option would be to add a Type field to ServiceClient and switch on it in the (not yet existing) ServiceClient Request method.

I think what's currently in [client.go](https://github.com/gophercloud/gophercloud/blob/master/openstack/client.go) is a good start. It has the standard service type names and allows a user to override a service type name. I don't know of an application currently taking advantage of that, _but_ the mechanism is there.

I think the `EndpointLocator` type is a good place to implement the first referenced doc without too much trouble. If the `EndpointLocator` functions implemented the whole spec, there would be a good amount of logic added to them, including reading a possibly bundled external spec sheet of service types and their known aliases. But one step at a time.

Now, an interesting thing mentioned in the second doc is how the version header is specified:

```
OpenStack-API-Version: [SERVICE_TYPE] [VERSION_STRING]
```

This is different than what's currently implemented in Gophercloud and what has been proposed for Manila. We should see if the above format is valid or if it's just a future spec.

If it is a valid format, this makes things a little easy: the `type` is already being declared when creating the service client, so as mentioned initially, add `Type` to the service client and set it, then use it to form the standardized URL.

But if URLs like `X-OpenStack-Nova-API-Version` and `X-Openstack-Manila-Api-Version` are the only accepted forms at this time, then Gophercloud should probably do the same pattern as it's doing with type: apply a sane default, but allow it to be overridden. 

Ideally the header _and_ requested version should be easily configurable to the user, so maybe extend `EndpointOpts` with these?

And while typing this out, it's probably a good idea to just allow the header to be overridden anyway, just in case.

Thoughts?

**Comment by pospispa** [comment]
@jtopjian thank you for you thoughts and links.
I plan to have look into it tomorrow.
I created PR https://github.com/gophercloud/gophercloud/pull/346 to give you an idea how I would like to use Manila microversion in Grant Access request that is supported from 2.7+ microversion.

**Comment by pospispa** [comment]
IMHO, it's necessary to extend the `ServiceClient` struct with extra data containing valid values describing the particular service endpoint as:
- Service type.
- Microversion.
- Service name.
- Other data available in service catalog, particularly [v2token](https://github.com/gophercloud/gophercloud/blob/master/openstack/identity/v2/tokens/results.go#L48) and [v3token](https://github.com/gophercloud/gophercloud/blob/master/openstack/identity/v3/tokens/results.go#L13).

The data are available in the `V2EndpointURL` and `V3EndpointURL` functions, however, these functions return only string containing endpoint URL. All other information are "forgotten".

That's why I tried to modify the `V2EndpointURL` and `V3EndpointURL` functions in such a way that they will return `EndpointInfo` struct instead of string [here](https://github.com/pospispa/gophercloud/tree/EndpointLocator-func-returns-EndpointInfo-instead-of-string). However, this approach does not seem feasible to me so far.

IMO, the `V2EndpointURL` and `V3EndpointURL` functions implement [the first document](http://docs-draft.openstack.org/15/462815/4/check/gate-api-wg-docs-ubuntu-xenial/b8479be//doc/build/html/guidelines/consuming-catalog.html) correctly.

I'm not able to find out another way how to propagate service type (and other needed information) from the `V2EndpointURL` and `V3EndpointURL` functions to the `ServiceClient` struct except changing the `EndpointLocator` type return type. Maybe changing the type from
```
type EndpointLocator func(EndpointOpts) (string, error)
```
to
```
type EndpointLocator func(EndpointOpts) (string, EndpointExtraInfo, error)
```

**Comment by pospispa** [comment]
I tried the version header from the [second doc](https://github.com/openstack/api-wg/blob/master/guidelines/microversion_specification.rst):
```
OpenStack-API-Version: [SERVICE_TYPE] [VERSION_STRING]
```
however, it did not work.

In addition, I searched for the `OpenStack-API-Version` string in the [Manila source code](https://github.com/openstack/manila), but did not find any occurrence. I also searched in this repo for the `X-Openstack-Manila-Api-Version` string and found 15 occurrences.

So it seems that headers like `X-OpenStack-Nova-API-Version` and `X-Openstack-Manila-Api-Version` are currently the only accepted forms. 
However, [the map](https://github.com/pospispa/gophercloud/blob/microversion-header-fix/service_client.go#L42-L49) proposed in the https://github.com/gophercloud/gophercloud/pull/341 can be replaced by a func like:
```
func getHTTPMicroversionHeader(serviceType, microversion string) (httpHeader string, value string)
```
in order to have the implementation more generic and ready for future specs.

**Comment by pospispa** [comment]
@jtopjian would you, please, give me feedback on my approach?

**Comment by jtopjian** [comment]
Yes, I'll try to look at it in the coming days. I've tried a few times previously but realized I need to set aside a few hours to look into all of this. Sorry for the delay.

**Comment by jtopjian** [comment]
I just reviewed #355. It allowed me to realize that I was overloading this feature with Version Discovery, which is probably out of scope here.

> IMO, the V2EndpointURL and V3EndpointURL functions implement the first document correctly.

Yes, you're correct.

What I was mixing in was Version Discovery: parsing the `http://example.com:port/` responses for the supported versions. My initial thought was to mix version discovery in with the EndpointURL functions, but maybe that's best done elsewhere? I can see it possibly happening somewhere in the `initClientOpts` function.

Thinking out loud: it's probably desirable for the application to be able to detect what versions are available so it could react accordingly. For example, let's say `client.Microversion` was set to the latest detected version by Gophercloud:

```go
if novaClient.Microversion == "2.42" {
  // The nova images proxy API is no longer supported in 2.42
  if novaClient.MinimumMicroversion == "2.1" {
    v := novaClient.Microversion
    novaClient.Microversion = "2.1"
    allPages := novaimages.ListDetail(novaClient, nil).AllPages()
    ...
    novaClient.Microversion = v
  }
}
```

It's a contrived example, but I think there's some value to it. But, I do think that it's out of scope here.

**Comment by jrperritt** [comment]
Closed by #355
